File: src/scripts/omnikey/values.ts
/** @noSelfInFile **/

import { PrefabCopy } from "../../modmain";

export function weaponValue(item: PrefabCopy) {
  return valueByDamage(item);
}

export function axeValue(item: PrefabCopy) {
  return toolValue(item, GLOBAL.ACTIONS.CHOP);
}

export function pickaxeValue(item: PrefabCopy) {
  return toolValue(item, GLOBAL.ACTIONS.MINE);
}

export function clothHeadValue(item: PrefabCopy) {
  return clothValue(item, GLOBAL.EQUIPSLOTS.HEAD);
}

export function clothBodyValue(item: PrefabCopy) {
  return clothValue(item, GLOBAL.EQUIPSLOTS.BODY);
}

export function armorHeadValue(item: PrefabCopy) {
  return armorValue(item, GLOBAL.EQUIPSLOTS.HEAD);
}

export function armorBodyValue(item: PrefabCopy) {
  return armorValue(item, GLOBAL.EQUIPSLOTS.BODY);
}

export function lightValue(item: PrefabCopy) {
  const index = [
    "torch",
    "lighter",
    "tarlamp",
    "minerhat",
    "lantern",
    "bottlelantern",
    "molehat",
    "gears_hat_goggles",
  ].indexOf(item.prefab);
  if (index === -1) return 0;
  return 1 / index;
}

export function caneValue(item: PrefabCopy) {
  return (
    (item.components.equippable &&
      item.components.equippable.walkspeedmult &&
      item.components.equippable.walkspeedmult > 1 &&
      item.components.equippable.walkspeedmult) ||
    0
  );
}

export function clothValue(item: PrefabCopy, slot: GLOBAL.EQUIPSLOTS) {
  return (
    (canBeEquipped(item, slot) &&
      valueByInsulation(item) + valueByDapperness(item) / 10) ||
    0
  );
}

export function armorValue(item: PrefabCopy, slot: GLOBAL.EQUIPSLOTS) {
  return (
    (canBeEquipped(item, slot) &&
      valueByArmor(item) + clothValue(item, slot)) ||
    0
  );
}

export function toolValue(item: PrefabCopy, action: any) {
  return (
    (item.components.tool !== undefined &&
      item.components.tool.actions[action] &&
      valueByConsumption(item, action) * 1000 + valueByUsage(item)) ||
    0
  );
}

function valueByDapperness(item: PrefabCopy) {
  if (!item.components.equippable) return 0;
  return item.components.equippable.dapperness || 0;
}

function valueByInsulation(item: PrefabCopy) {
  if (!item.components.insulator) return 0;
  return item.components.insulator.insulation || 0;
}

function canBeEquipped(item: PrefabCopy, slot: any) {
  return (
    item.components.equippable && item.components.equippable.equipslot === slot
  );
}

function valueByDamage(item: PrefabCopy) {
  return (
    (item.components.weapon &&
      item.components.weapon.damage * 100 + valueByUsage(item)) ||
    0
  );
}

function valueByUsage(item: PrefabCopy): number {
  return (
    (item.components.finiteuses !== undefined &&
      item.components.finiteuses.total * 10 -
        item.components.finiteuses.current) ||
    0
  );
}

function valueByConsumption(item: PrefabCopy, action: any): number {
  return (
    (item.components.finiteuses !== undefined &&
      1 / item.components.finiteuses.consumption[action]) ||
    100
  );
}

function valueByArmor(item: PrefabCopy) {
  return (
    (item.components.armor &&
      item.components.armor.absorb_percent * 100000 -
        item.components.armor.condition) ||
    0
  );
}

File: scripts/omnikey/values.lua
local exports = {}

local function valueByUsage(item)
	return item.components.finiteuses ~= nil
			and item.components.finiteuses.total * 10 - item.components.finiteuses.current
		or 0
end

local function valueByDapperness(item)
	if not item.components.equippable then
		return 0
	end
	return item.components.equippable.dapperness or 0
end

local function valueByInsulation(item)
	if not item.components.insulator then
		return 0
	end
	return item.components.insulator.insulation or 0
end

local function canBeEquipped(item, slot)
	return item.components.equippable and item.components.equippable.equipslot == slot
end

local function valueByDamage(item)
	return item.components.weapon and item.components.weapon.damage * 100 + valueByUsage(item) or 0
end

local function valueByConsumption(item, action)
	return item.components.finiteuses ~= nil and 1 / item.components.finiteuses.consumption[action] or 100
end

local function valueByArmor(item)
	return item.components.armor and item.components.armor.absorb_percent * 100000 - item.components.armor.condition
		or 0
end

local function valueByToolAction(item, action)
	return item.components.tool
			and item.components.tool.actions[action]
			and valueByConsumption(item, action) * 1000 + valueByUsage(item)
		or 0
end

---@param item ds.entity
---@param slot ds.equipslot
local function valueByClothEquipSlot(item, slot)
	return canBeEquipped(item, slot) and valueByInsulation(item) + valueByDapperness(item) / 10 or 0
end

---@param item ds.entity
---@param slot ds.equipslot
local function valueByArmorEquipSlot(item, slot)
	return canBeEquipped(item, slot) and valueByArmor(item) + valueByClothEquipSlot(item, slot) or 0
end

---@param action ds.action
function exports.toolValue(action)
	---@param item ds.entity
	return function(item)
		return valueByToolAction(item, action)
	end
end

---@param item ds.entity
function exports.weaponValue(item)
	return valueByDamage(item)
end

---@param slot ds.equipslot
function exports.clothValue(slot)
	---@param item ds.entity
	return function(item)
		return valueByClothEquipSlot(item, slot)
	end
end

---@param slot ds.equipslot
function exports.armorValue(slot)
	---@param item ds.entity
	return function(item)
		return valueByArmorEquipSlot(item, slot)
	end
end

---@param item ds.entity
function exports.lightValue(item)
	for i, v in ipairs({
		"torch",
		"lighter",
		"tarlamp",
		"minerhat",
		"lantern",
		"bottlelantern",
		"molehat",
		"gears_hat_goggles",
	}) do
		if item.prefab == v then
			return 1 / i
		end
	end
	return 0
end

---@param item ds.entity
function exports.caneValue(item)
	return item.components.equippable
			and item.components.equippable.walkspeedmult
			and item.components.equippable.walkspeedmult > 1
			and item.components.equippable.walkspeedmult
		or 0
end

---@param item ds.entity
function exports.weaponValuePrioritizeHambat(item)
	return item.prefab == "hambat" and 10000 or exports.weaponValue(item)
end

return exports
